### PR TITLE
Bug 1780806 - Allow the uploader to signal it is done handling uploads.

### PR DESCRIPTION
**Pull request by badboy**
Putting this up so it's up.
I want to discuss this with chutten when he's back.

I'm not super happy with it because the `UploadResult::Done` needs to be handled on a different layer than all the other `UploadResult` variants.
(It's also only implemented for the RLB right now)

### CODE REVIEW

File: glean-core/rlb/src/test.rs [hunk 1]
```diff
@@ -1131,3 +1135,73 @@ fn test_boolean_get_num_errors() {
...
+    // Give it time to spin down.
+    thread::sleep(Duration::from_millis(UPLOAD_WAIT_TIME * 2));
+
+    // Submit another ping and wait for it to do work.
+    custom_ping.submit(None);
+    thread::sleep(Duration::from_millis(UPLOAD_WAIT_TIME * 2));
+
+    let map = call_count.lock().unwrap();
+    assert_eq!(2, map.len(), "should have launched 2 uploader threads");
+    for (_tid, &count) in &*map {
```
Comment (chutten) [inline glean-core/rlb/src/test.rs:1204]: How many threads do we use to call out to the uploader, anyway? If it's one per submission, then this won't properly ensure we've handled `Done` properly

File: glean-core/rlb/src/net/mod.rs [hunk 2]
```diff
@@ -97,21 +98,35 @@ impl UploadManager {
                             let upload_url = format!("{}{}", inner.server_endpoint, request.path);
                             let headers: Vec<(String, String)> =
                                 request.headers.into_iter().collect();
-                            let result = inner.uploader.upload(upload_url, request.body, headers);
+                            let result =
+                                match inner.uploader.upload(upload_url, request.body, headers) {
```
Comment (chutten) [inline glean-core/rlb/src/net/mod.rs:102]: If you're concerned about maintaining the abstraction, you could drill through a return value from `PingUploadManager::process_ping_upload_response`. Then you could add this `UploadResult::Done` case to its `match`, returning `PingUploadTask::Done` or something. Then handle the correct `Done` here.

Comment (badboy) [reply glean-core/rlb/src/net/mod.rs:102]: I'm not sure I understand what you mean.
Right now `glean_process_ping_upload_response` doesn't return anything. Are you suggesting it _should_ return something that this uploading loop handles to determine the next step?
(That again doesn't seem _too_ bad I think)

**Comment by badboy** [comment]
Ok, I reworked this a bit now.
The `Done` is passed through and then `glean_process_ping_upload_response` returns something that instructs the upload loop whether to continue or end.  I also implemented that for Kotlin now to show that it works.

Additionally I reworked the test. Instead of arbitrary sleep times it now uses a barrier to sync up the 2 threads. That should be much more reliable.

File: glean-core/rlb/src/net/mod.rs (hunk 2 above)
Comment (chutten) [reply glean-core/rlb/src/net/mod.rs]: That's precisely what I meant. Have the thing that handles `UploadResult::*` handle this new one, then have it return the result all the way back out here.

File: glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt [hunk 3]
```diff
@@ -112,13 +113,19 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
                     )
 
                     // Process the upload response
-                    gleanProcessPingUploadResponse(action.request.documentId, result)
+                    val action = gleanProcessPingUploadResponse(action.request.documentId, result)
+                    when (action) {
+                        UploadTaskAction.CONTINUE -> continue
+                        UploadTaskAction.END -> break
+                    }
```
Comment (chutten) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:120]: No longer RLB-only, I see.

File: glean-core/rlb/src/net/mod.rs [hunk 4]
```diff
@@ -99,19 +100,25 @@ impl UploadManager {
...
                             let result = inner.uploader.upload(upload_url, request.body, headers);
                             // Process the upload response.
-                            glean_core::glean_process_ping_upload_response(doc_id, result);
+                            match glean_core::glean_process_ping_upload_response(doc_id, result) {
+                                UploadTaskAction::Continue => continue,
+                                UploadTaskAction::End => break,
+                            }
                         }
                         PingUploadTask::Wait { time } => {
+                            log::trace!("Instructed to wait for {:?}ms", time);
```
Comment (chutten) [inline glean-core/rlb/src/net/mod.rs:109]: If we're tracing the other `PingUploadTask::*` we should trace `PingUploadTask::Upload`, no?

File: .detekt.yml [hunk 5]
```diff
@@ -9,6 +9,14 @@ style:
   # Multiple returns are ok
   ReturnCount:
     active: false
+
+  # Go ahead and jump (Jump!)
+  # Go ahead and jump
```
Comment (chutten) [inline .detekt.yml:14]: 👏

File: glean-core/rlb/src/test.rs [hunk 6]
```diff
@@ -1131,3 +1134,80 @@ fn test_boolean_get_num_errors() {
...
+    let _t = new_glean(Some(cfg), true);
+
+    // Define a new ping and submit it.
+    const PING_NAME: &str = "test-ping";
+    let custom_ping = private::PingType::new(PING_NAME, true, true, vec![]);
+    custom_ping.submit(None);
+    custom_ping.submit(None);
+
+    // Sync up with the upload thread.
+    barrier.wait();
```
Comment (chutten) [inline glean-core/rlb/src/test.rs:1199]: Ah, and if the threading model changes (ie, if each `submit` starts getting its own thread) this test will fail by locking itself?

File: glean-core/src/upload/result.rs [hunk 7]
```diff
@@ -32,6 +32,14 @@ pub enum UploadResult {
         /// The HTTP status code
         code: i32,
     },
+
+    /// Signal that this uploader is done with work
+    /// and doesn't accept new work.
```
Comment (chutten) [inline glean-core/src/upload/result.rs:37]:
```suggestion
    /// and won't accept new work.
```

**Review by chutten — COMMENTED** [review]
Needs CHANGELOG

File: glean-core/rlb/src/test.rs (hunk 6 above)
Comment (badboy) [reply glean-core/rlb/src/test.rs:1199]: yep

Comment (badboy) [reply glean-core/rlb/src/test.rs:1199]: well, in theory. *goes back to the code* why doesn't this fail now? oO

Comment (badboy) [reply glean-core/rlb/src/test.rs:1199]: ugh, turns out I'm still running into timing issues here. Let me once again rethink the test.

**Review by chutten — APPROVED** [review]

**Pull request merged**